File: compiler/src/typechecker/symbolTable.ts
import { TypeAnnotation } from "../ast/nodes.js";
import { SymbolBinding, SymbolOrigin } from "./binder.js";

export interface SymbolInfo {
  type: TypeAnnotation;
  constant: boolean;
  binding?: SymbolBinding;
  origin?: SymbolOrigin;
  source?: string;
}

export class SymbolTable {
  private scopes: Map<string, SymbolInfo>[] = [new Map()];

  enterScope(): void {
    this.scopes.push(new Map());
  }

  exitScope(): void {
    if (this.scopes.length === 1) {
      throw new Error("Cannot exit the global scope.");
    }
    this.scopes.pop();
  }

  declare(name: string, info: SymbolInfo): boolean {
    const current = this.scopes[this.scopes.length - 1];
    if (!current) {
      throw new Error("No active scope.");
    }

    // Check if already declared in current scope
    if (current.has(name)) {
      return false;
    }

    current.set(name, info);
    return true;
  }

  lookup(name: string): SymbolInfo | undefined {
    // Search from innermost to outermost scope
    for (let i = this.scopes.length - 1; i >= 0; i--) {
      const symbol = this.scopes[i]?.get(name);
      if (symbol) {
        return symbol;
      }
    }
    return undefined;
  }

  has(name: string): boolean {
    return this.lookup(name) !== undefined;
  }

  isConstant(name: string): boolean {
    return this.lookup(name)?.constant ?? false;
  }


  getScopeDepth(): number {
    return this.scopes.length;
  }

  hasInCurrentScope(name: string): boolean {
    const current = this.scopes[this.scopes.length - 1];
    return current?.has(name) ?? false;
  }

  /** Every name visible right now, from every open scope, for typo suggestions. */
  allNames(): string[] {
    const names = new Set<string>();
    for (const scope of this.scopes) {
      for (const name of scope.keys()) names.add(name);
    }
    return [...names];
  }
}

File: crates/raven-core/src/symbol_table.rs
//! Scoped symbol table, mirroring `compiler/src/typechecker/symbolTable.ts`.

use std::collections::HashMap;

use crate::type_::RavenType;

use super::binder::SymbolOrigin;

/// A symbol entry, mirroring the TS `SymbolInfo`.
#[derive(Debug, Clone)]
pub struct SymbolInfo {
    pub type_: RavenType,
    pub constant: bool,
    pub binding: Option<usize>,
    pub origin: Option<SymbolOrigin>,
    pub source: Option<String>,
}

/// A stack of scoped symbol maps, mirroring `SymbolTable`.
#[derive(Debug)]
pub struct SymbolTable {
    scopes: Vec<HashMap<String, SymbolInfo>>,
}

impl Default for SymbolTable {
    fn default() -> Self {
        Self {
            scopes: vec![HashMap::new()],
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SymbolTableError {
    CannotExitGlobalScope,
    NoActiveScope,
}

impl SymbolTable {
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    pub fn enter_scope(&mut self) {
        self.scopes.push(HashMap::new());
    }

    pub fn exit_scope(&mut self) -> Result<(), SymbolTableError> {
        if self.scopes.len() == 1 {
            return Err(SymbolTableError::CannotExitGlobalScope);
        }
        self.scopes.pop();
        Ok(())
    }

    /// Declare in the current scope. Returns `false` if the name is already
    /// declared in the current scope.
    pub fn declare(&mut self, name: String, info: SymbolInfo) -> bool {
        let current = self
            .scopes
            .last_mut()
            .ok_or(SymbolTableError::NoActiveScope)
            .expect("scope stack is never empty");
        if current.contains_key(&name) {
            return false;
        }
        current.insert(name, info);
        true
    }

    pub fn lookup(&self, name: &str) -> Option<&SymbolInfo> {
        for scope in self.scopes.iter().rev() {
            if let Some(symbol) = scope.get(name) {
                return Some(symbol);
            }
        }
        None
    }

    #[must_use]
    pub fn has(&self, name: &str) -> bool {
        self.lookup(name).is_some()
    }

    #[must_use]
    pub fn is_constant(&self, name: &str) -> bool {
        self.lookup(name).is_some_and(|s| s.constant)
    }

    #[must_use]
    pub fn scope_depth(&self) -> usize {
        self.scopes.len()
    }

    #[must_use]
    pub fn has_in_current_scope(&self, name: &str) -> bool {
        self.scopes
            .last()
            .is_some_and(|scope| scope.contains_key(name))
    }

    /// All visible names across all scopes (deduplicated).
    #[must_use]
    pub fn all_names(&self) -> Vec<String> {
        let mut names = std::collections::BTreeSet::new();
        for scope in &self.scopes {
            for name in scope.keys() {
                names.insert(name.clone());
            }
        }
        names.into_iter().collect()
    }
}
